go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,195 @@
+	controllerStop <-chan struct{}
+}
+
+type dnsEntry struct {
+	//this map holds all the namespaces that a dnsName appears in
+	namespaces map[string]struct{}
+	// the current IP addresses the dnsName resolves to
+	dnsResolves []net.IP
+	// the addressSet that contains the current IPs
+	dnsAddressSet AddressSet
+}
+
+func NewEgressDNS(addressSetFactory AddressSetFactory, controllerStop <-chan struct{}) (*EgressDNS, error) {
+	dnsInfo, err := util.NewDNS("/etc/resolv.conf")
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil, err
+	}
+
+	egressDNS := &EgressDNS{
+		dns:               dnsInfo,
+		dnsEntries:        make(map[string]*dnsEntry),
+		addressSetFactory: addressSetFactory,
+
+		added:          make(chan bool),
+		stopChan:       make(chan struct{}),
+		controllerStop: controllerStop,
+	}
+
+	go utilwait.Until(egressDNS.Sync, 0, egressDNS.stopChan)
+
+	return egressDNS, nil
+}
+
+func (e *EgressDNS) Add(namespace, dnsName string) (AddressSet, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if _, exists := e.dnsEntries[dnsName]; !exists {
+		var err error
+		dnsEntry := dnsEntry{
+			namespaces: make(map[string]struct{}),
+		}
+		if e.addressSetFactory == nil {
+			panic("addressSetFactory is null, that cannot happen")
+		}
+		dnsEntry.dnsAddressSet, err = e.addressSetFactory.NewAddressSet(dnsName, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create addressSet for %s: %v", dnsName, err)
+		}
+		e.dnsEntries[dnsName] = &dnsEntry
+		//only call Add if the dnsName doesn't exist in dnsEntries
+		if err := e.dns.Add(dnsName); err != nil {
+			klog.Errorf("DNS Name: %s cannot be added to the dnsMap - %v", dnsName, err)
+		}
+		unsorted := e.dns.GetIPs(dnsName)
+		e.dnsEntries[dnsName].dnsResolves = SortIPs(unsorted)
+		err = e.dnsEntries[dnsName].dnsAddressSet.AddIPs(e.dnsEntries[dnsName].dnsResolves)
+		if err != nil {
+			return nil, fmt.Errorf("cannot add IPs to EgressFirewalls AddressSet: %s - %v", dnsName, err)
+		}
+		e.signalAdded()
+	}
+	e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+	return e.dnsEntries[dnsName].dnsAddressSet, nil
+
+}
+
+func (e *EgressDNS) Delete(namespace string) bool {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	// go through all dnsNames for namespaces
+	for dnsName, dnsEntry := range e.dnsEntries {
+		// delete the dnsEntry
+		delete(dnsEntry.namespaces, namespace)
+		if len(dnsEntry.namespaces) == 0 {
+			// the dnsEntry appears in no other namespace so delete the address_set
+			err := dnsEntry.dnsAddressSet.Destroy()
+			if err != nil {
+				klog.Errorf("Error deleteing EgressFirewall AddressSet for dnsName: %s %v", dnsName, err)
+			}
+			// the dnsEntry is no longer needed because nothing referances it delete it
+			delete(e.dnsEntries, dnsName)
+		}
+	}
+	return len(e.dnsEntries) == 0
+}
+
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	return e.dns.Update(dns)
+}
+
+func (e *EgressDNS) Sync() {
```

Comment:
It would be good to have a bit of docblock explaining what this does; this is a lot of high-context code.